pkg/hub/addon/discovery_controller.go

```diff
@@ -0,0 +1,246 @@
+}
+
+func (c *addOnFeatureDiscoveryController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	// The value of queueKey might be
+	// 1) equal to the default queuekey. It is triggered by resync every 10 minutes;
+	// 2) in format: namespace/name. It indicates the event source is a ManagedClusterAddOn;
+	// 3) in format: name. It indicates the event source is a ManagedCluster;
+	queueKey := syncCtx.QueueKey()
+	namespace, name, err := cache.SplitMetaNamespaceKey(queueKey)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil
+	}
+
+	switch {
+	case queueKey == factory.DefaultQueueKey:
+		// handle resync
+		return c.resync(syncCtx)
+	case len(namespace) > 0:
+		// sync a particular addon
+		return c.syncAddOn(ctx, namespace, name)
+	default:
+		// sync the cluster
+		return c.syncCluster(ctx, name)
+	}
+}
+
+func (c *addOnFeatureDiscoveryController) resync(syncCtx factory.SyncContext) error {
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, cluster := range clusters {
+		if !cluster.DeletionTimestamp.IsZero() {
+			continue
+		}
+		syncCtx.Queue().Add(cluster.Name)
+	}
+	return nil
+}
+
+func (c *addOnFeatureDiscoveryController) syncAddOn(ctx context.Context, clusterName, addOnName string) error {
+	klog.V(4).Infof("Reconciling addOn %q", addOnName)
+
+	addOn, err := c.addOnLister.ManagedClusterAddOns(clusterName).Get(addOnName)
+	if errors.IsNotFound(err) {
+		// addon is deleted
+		return c.deleteAddOnLabel(ctx, clusterName, addOnName)
+	}
+	if err != nil {
+		return err
+	}
+
+	// addon is deleting
+	if !addOn.DeletionTimestamp.IsZero() {
+		return c.deleteAddOnLabel(ctx, clusterName, addOnName)
+	}
+
+	return c.updateAddOnLabel(ctx, clusterName, addOn)
+}
+
+func (c *addOnFeatureDiscoveryController) syncCluster(ctx context.Context, clusterName string) error {
+	// sync all addon labels on the managed cluster
+	cluster, err := c.clusterLister.Get(clusterName)
+	if errors.IsNotFound(err) {
+		// cluster is deleted
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Do not update addon label if cluster is deleting
+	if !cluster.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	newLabels := map[string]string{}
+	for key, value := range cluster.Labels {
+		if strings.HasPrefix(key, addOnFeaturePrefix) {
+			continue
+		}
+		newLabels[key] = value
+	}
+	addOns, err := c.addOnLister.ManagedClusterAddOns(clusterName).List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("unable to list addOns of cluster %q: %w", clusterName, err)
+	}
+	for _, addOn := range addOns {
+		// addon is deleting
+		if !addOn.DeletionTimestamp.IsZero() {
+			continue
+		}
+		key := addOnFeaturePrefix + addOn.Name
+		newLabels[key] = getAddOnLabelValue(addOn)
+	}
+
+	// no work if the cluster labels have no change
+	if (len(newLabels) == 0 && len(cluster.Labels) == 0) || reflect.DeepEqual(newLabels, cluster.Labels) {
```

Comment:
nit: if you  use apiequality.sementic.DeepEqual you would not need to compare the length